presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/README.md

```diff
@@ -0,0 +1,162 @@
+
+### Supporting new PlanNodes
+
+There are several things you might want to verify in a PlanNode:
+
+1. The contents of the PlanNode itself (e.g. equi criteria in a JoinNode)
+2. The contents of some Symbol-producing member of the PlanNode (e.g.
+   FunctionCalls in a WindowNode)
+
+These should be handled separately.
+
+#### Fields that do not produce Symbols
+
+Fields that don't produce symbols should be verified in a new Matcher that
+is specific to the PlanNode. As much as possible, the matcher should call
+an object's own .equals method rather than implementing the comparison itself.
+
+See JoinMatcher for an example of comparing non-symbol producing fields.
+
+#### Fields that produce Symbols
+
+Fields that produce Symbols should be compared with an RvalueMatcher that can
+be used with the special `Matcher` `Alias`. `Alias` is used to bind those
+Symbols to an alias so they can be referenced by alias further up the tree. 
+
+Examples of matching fields that produce symbols:
+
+Node type | Matching infrastructure
+----------|------------------------
+TableScanNode | ColumnReference
+ProjectNode | ExpressionAssignment
+AggregationNode | AggregationFunctionMatcher and FunctionCallProvider 
+
+## Aliases and ExpectedValueProviders
+
+At a high level, Presto plans represent data through Symbols. A TableScanNode,
+for instance, generates one symbol per column in the table as its output.
+ProjectNodes can apply a calculation to some input Symbol and generate a new
+Symbol as output.
+
+If you are writing tests, you may care about what is happening to data in the
+actual plan. The simplest possible example of this would be asserting that a
+`SELECT name FROM people` query has an OutputNode with an output symbol that
+actually represents the data in the `name` column of the `people` table.
+
+### Aliases
+
+In order for the plan testing framework to be independent of how Presto
+allocates Symbols, Symbols must be given an alias where they originate, and
+must be referenced by that alias further up the plan.
+
+Binding the alias to the Symbol where it originates must also be independent of
+the SymbolAllocator. This means that the binding must match the right side of
+the assignment; i.e. the value that is being assigned to the symbol.
+
+#### Alias binding example
+
+As an example, take a TableScanNode:
+
+```
+  TableScanNode t {
+    table: TableHandle("people")
+    assignments: { symbol_1: ColumnHandle("name") }
+  }
+```
+
+We can match the value Presto assigns to `symbol_1` by matching a ColumnHandle
+for the `name` column  in the `people` table using `new ColumnReference("people",
+"name")`
+
+Actually creating a binding is done using the special `Matcher` `Alias`.
+
+```
+  Matcher matcher = new Alias("NAME_ALIAS", new ColumnReference("people", "name");
+```
+
+This `Matcher` can then be added to an existing `PlanMatchPattern` using
+`with()`. This would be cumbersome. In practice, any non-trivial node should
```

Comment:
I don't think it's really necessary (part of how to write a `static` method).
It's too low level detail while we can catch that on review if one does not express his Alias Matching clearly.